### PR TITLE
bpo-20028: Empty escapechar/quotechar is not allowed for csv.Dialect

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue20028

### CODE REVIEW

File: Modules/_csv.c [hunk 1]
```diff
@@ -250,7 +250,7 @@ _set_char_or_none(const char *name, Py_UCS4 *target, PyObject *src, Py_UCS4 dflt
             if (len < 0) {
                 return -1;
             }
-            if (len > 1) {
+            if (len != 1) {
```
Comment (serhiy-storchaka) [inline Modules/_csv.c:253]: `if (len > 0)` below is no longer needed.

File: Modules/_csv.c [hunk 2]
```diff
@@ -283,7 +283,7 @@ _set_char(const char *name, Py_UCS4 *target, PyObject *src, Py_UCS4 dflt)
         if (len < 0) {
             return -1;
         }
-        if (len > 1) {
+        if (len != 1) {
```
Comment (serhiy-storchaka) [inline Modules/_csv.c:283]: `if (len > 0)` below is no longer needed. `*target = NOT_SET` above is no longer needed.

File: Modules/_csv.c [hunk 3]
```diff
@@ -250,7 +250,7 @@ _set_char_or_none(const char *name, Py_UCS4 *target, PyObject *src, Py_UCS4 dflt
             if (len < 0) {
                 return -1;
             }
-            if (len > 1) {
+            if (len != 1) {
                 PyErr_Format(PyExc_TypeError,
```
Comment (corona10) [inline Modules/_csv.c:254]: @serhiy-storchaka 
By the way, ValueError looks proper..

File: Doc/library/csv.rst [hunk 4]
```diff
@@ -402,6 +404,8 @@ Dialects support the following attributes:
    as the *delimiter* or *quotechar*, or which contain new-line characters.  It
    defaults to ``'"'``.
 
+   .. versionchanged:: 3.11
+      An empty *quotechar* is not allowed.
```
Comment (serhiy-storchaka) [inline Doc/library/csv.rst:408]: Was it allowed before?

File: Modules/_csv.c (hunk 3 above)
Comment (serhiy-storchaka) [reply Modules/_csv.c:254]: But TypeError looks proper too if consider a character (a string with length 1) a separate type. Traditionally TypeError is raised if 1-char string is expected but a string with different length is passed. For example
```
>>> ord('abc')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: ord() expected a character, but string of length 3 found
```

File: Lib/test/test_csv.py [hunk 5]
```diff
@@ -342,7 +342,8 @@ def test_read_escape(self):
         self._read_test(['a,^b,c'], [['a', 'b', 'c']], escapechar='^')
         self._read_test(['a,\0b,c'], [['a', 'b', 'c']], escapechar='\0')
         self._read_test(['a,\\b,c'], [['a', '\\b', 'c']], escapechar=None)
-        self._read_test(['a,\\b,c'], [['a', '\\b', 'c']], escapechar='')
+        self.assertRaises(TypeError, self._read_test,
+                          ['a,\\b,c'], [['a', '\\b', 'c']], escapechar='')
```
Comment (serhiy-storchaka) [inline Lib/test/test_csv.py]: Invalid values should be tested in different method, in `_test_arg_valid`.

File: Modules/_csv.c (hunk 1 above)
Comment (corona10) [reply Modules/_csv.c:253]: Thanks nice catch!

File: Doc/library/csv.rst (hunk 4 above)
Comment (corona10) [reply Doc/library/csv.rst:408]: ## AS-IS
`````
Python 3.9.1 (v3.9.1:1e5d33e9b9, Dec  7 2020, 12:10:52)
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import csv
>>> class mydialect(csv.Dialect):
...     delimiter = ";"
...     escapechar = '\\'
...     doublequote = False
...     skipinitialspace = True
...     lineterminator = '\r\n'
...     quoting = csv.QUOTE_NONE
...
>>> mydialect.quotechar = ""
>>> mydialect()
<__main__.mydialect object at 0x7fa336232f10>
`````

## TO-BE
`````
Python 3.11.0a1+ (heads/bpo-20028-zero-length:d8f1230f5e, Oct 10 2021, 01:30:52) [Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import csv
>>> class mydialect(csv.Dialect):
...     delimiter = ";"
...     escapechar = '\\'
...     doublequote = False
...     skipinitialspace = True
...     lineterminator = '\r\n'
...     quoting = csv.QUOTE_NONE
...
>>> mydialect.quotechar = ""
>>> mydialect()
Traceback (most recent call last):
  File "/Users/user/oss/cpython/Lib/csv.py", line 49, in _validate
    _Dialect(self)
    ^^^^^^^^^^^^^^
TypeError: "quotechar" must be a 1-character string

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/user/oss/cpython/Lib/csv.py", line 45, in __init__
    self._validate()
    ^^^^^^^^^^^^^^^^
  File "/Users/user/oss/cpython/Lib/csv.py", line 52, in _validate
    raise Error(str(e))
    ^^^^^^^^^^^^^^^^^^^
_csv.Error: "quotechar" must be a 1-character string
`````

File: Lib/test/test_csv.py (hunk 5 above)
Comment (corona10) [reply Lib/test/test_csv.py]: Please see other cases :)
https://github.com/python/cpython/blob/b454e8e4df73bc73bc1a6f597431f171bfae8abd/Lib/test/test_csv.py#L318

_test_arg_valid only tests csv.Reader/Writer currently.

Comment (serhiy-storchaka) [reply Lib/test/test_csv.py]: That test is a test for invalid input.

`_test_arg_valid` is a test for invalid options. TypeError for `escapechar=''` does not depend on input.

File: Doc/library/csv.rst (hunk 4 above)
Comment (serhiy-storchaka) [reply Doc/library/csv.rst:408]: Ah, so it was allowed with `quoting=QUOTE_NONE`. It would be worth to add a test for this case in `_test_arg_valid`.

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**